vrs/marten-construction-script.py

```diff
@@ -0,0 +1,230 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# Daniel Marten 01-18-2022
+# Construct a ~100k Variant Test Database for GA4GH-VRS Annotation Script
+# ran in broad-mpg-gnomad/gnomad_marten as marten-assembly-v3.ipynb
+# major changes from last version: used Black to format, added temporary and final path variables, 
+# added variables for number of variants for each type, de-duplicated, and fixed the header!
+
+import hail as hl
+import random
+from random import sample
+from gnomad.resources.grch38.gnomad import public_release
+import logging
+
+# Setting randomness
+random.seed(505)
+hl.init(global_seed=5)
+
+# Import the gnomAD v3.1.2 full variant table (approx 759,000,000 variants)
+# note: no need to repartition the whole table, as it was recently repartitioned to 10,000
+
+ht_whole = public_release("genomes").ht()
+
+"""
+PREVIOUSLY CALCULATED FREQUENCY STATISTICS
+
+Frequency of the following variant types in gnomAD v3:
+- indels
+- long reference alleles (>5bp; long reference)
+- long alternate alleles (>5bp; long variant)
+- X chromosome variants 
+- Y chromosome variants
+
+Calculated with the following code:
+whole_count = ht_whole.count()
+indel_freq = ht_whole.aggregate(hl.agg.count_where(hl.is_indel(ht_whole.alleles[0], ht_whole.alleles[1]))) / whole_count
+ref_freq = ht_whole.aggregate(hl.agg.count_where(hl.len(ht_whole.alleles[0])>5)) / whole_count
+var_freq = ht_whole.aggregate(hl.agg.count_where(hl.len(ht_whole.alleles[1])>5)) / whole_count
+x_freq = ht_whole.aggregate(hl.agg.count_where(ht_whole.locus.contig == "chrX")) / whole_count
+y_freq = ht_whole.aggregate(hl.agg.count_where(ht_whole.locus.contig == "chrY")) / whole_count
+
+print(f'Frequencies as: indels {indel_freq} and long ref {ref_freq} and long var {var_freq} and x freq {x_freq} and y freq {y_freq}')
+
+Quite time-intensive, so ran previously to save computation headache
+
+Feel free to run yourself - at your own peril ! 
+"""
+
+indel_freq = 0.14525709561723196
+ref_freq = 0.02122722096390106
+var_freq = 0.0381956465303165
+x_freq = 0.0398928560027597
+y_freq = 0.0015374206699161612
+whole_count = 759302267
+
+# e.g.: for 1000 variants, 145 will be indels
+# for 1000 variants, 40 will be X chromosomes and only 1.5 will be Y chromosome
+
+# Set a temporary path to make writing checkpoints easier (and more easy to edit for future users)
+tmp_path = "gs://gnomad-tmp-4day/marten-temp-01-18-v3/"
+
+# Construct tables of samples from this
+# Makeup: 50k random, 10k extra indels, 10k more long references, 10k long variants, 10k sex chromosome
+# --> (9k X, 1k Y)
+
+# 50,000 random variants
+rand_variants = 50000
+randSampleFreq = rand_variants / whole_count
+ht_rand = ht_whole.sample(randSampleFreq)
+# ht_rand = ht_rand.naive_coalesce(100)
+# TO ADD: ht_50k = ht_50k.repartition(100)
+rand_path = tmp_path + "ht_rand_temp.ht"
+ht_rand = ht_rand.checkpoint(rand_path, _read_if_exists=True, overwrite=True)
+
+# 10,000 indels
```

Comment:
The section for creating each of the subset hts is a little repetitive and has a lot of variables. I think it can be simplified and would be easier to read if you store some of the information in dictionaries, something like this
(assuming ht_rand is already stored as union_ht) 
```suggestion
# Create freq dictionary with variant type as the key and pre-calculated frequencies as the value
freq_dict = {"indel": 0.14525709561723196,
"long_ref": 0.02122722096390106,
"long_var": 0.0381956465303165,
"on_x":  0.0398928560027597,
"on_y": 0.0015374206699161612}

# Create desired counts dictionary with variant type as the key and desired number of that variant type in the test subset as the value
desired_counts = {"indel": 10000,
"long_ref": 10000,
"long_var": 10000,
"on_x":  9000,
"on_y": 1000}

ht_whole = ht_whole.annotate(indel = hl.is_indel(ht_whole.alleles[0], ht_whole.alleles[1]),
   long_ref = hl.len(ht_whole.alleles[0]) > 5,
   long_var = hl.len(ht_whole.alleles[1]) > 5,
   on_x = (ht_whole.locus.contig == "chrX"),
   on_y = (ht_whole.locus.contig == "chrY")
   )

for variant_type in freq_dict:
    variant_freq = freq_dict[variant_type]
    desired_count = desired_counts[variant_type]
    sampling_freq = desired_count / (whole_count * variant_freq)
    
    ht = ht_whole.filter(ht_whole[variant_type])
    ht = ht.sample(sampling_freq)
    ht = ht.checkpoint(f'{tmp_path}{variant_type}_temp.ht', _read_if_exists=True, overwrite=True)
    union_ht = union_ht.union(ht)
```